File: grind75_python/MinimumWindowSubstring.py
# https://leetcode.com/problems/minimum-window-substring/

class Solution:
  def minWindow(self, s: str, t: str) -> str:
    ans = [-1,-1]
    remaining = len(t)
    l = 0
    r = 0
    dic = {}
    for char in t:
      if char in dic:
        dic[char] += 1
      else:
        dic[char] = 1

    while r<len(s):
      curr = s[r]
      if curr in dic:
        dic[curr]-=1
        if dic[curr]>=0:
          remaining-=1
        while remaining==0:
          if s[l] in dic:
            if ans[0]==ans[1]==-1 or ans[1]-ans[0]>r-l:
              ans = [l, r]
            dic[s[l]]+=1
            if dic[s[l]]>0:
              remaining+=1
          l+=1
      r+=1
    
    return "" if ans[0]==ans[1]==-1 else s[ans[0]:ans[1]+1]
  
s = Solution()
print(s.minWindow(s = "a", t = "a"))


        

File: CPP/minimumWindowSubstring.cpp
// https://leetcode.com/problems/minimum-window-substring/

#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  string minWindow(string s, string t) {
    unordered_map<char, pair<int, int>> present;  // character /no. in t / no. in substr
    int remaining = 0;
    int min = INT_MAX;
    int minfront, minrear;
    for (auto c : t) {
      if (present.find(c) == present.end()) {
        remaining++;
        present[c] = {0, 0};
      }
      present[c].first++;
    }

    int front = 0, rear = 0;
    while (present.find(s[rear]) == present.end() and rear<s.size()) {
      rear++;
      front++;
    }

    while (front < s.size()) {
      if (present.find(s[front]) != present.end()) {
        present[s[front]].second++;
        if (present[s[front]].second == present[s[front]].first) {
          remaining--;
        }
        // while (present[q.front().first].second > present[q.front().first].first) {
        //   present[q.front().first].second--;
        //   q.pop();
        //   rear = q.front().second;
        // }
        while (present.find(s[rear]) == present.end() or present[s[rear]].second > present[s[rear]].first) {
          if (present.find(s[rear]) != present.end()) {
            present[s[rear]].second--;
          }
          rear++;
        }
        if (min > front - rear + 1 and remaining == 0) {
          minfront = front;
          minrear = rear;
          min = front - rear + 1;
        }
      }
      front++;
    }

    return remaining ? "" : string(&s[minrear], &s[minfront + 1]);
  }
};

int main() {
  Solution s;
  cout << s.minWindow("a", "b") << endl;
}